### PR TITLE
Add Dremio and Dremio cross engine test

### PR DESCRIPTION
**Pull request by sagar-agarwal2404**
- add IcebergDremioExtension
- add DremioHelper
- add Dremio Section in the README

### CODE REVIEW

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+package org.projectnessie.integtests.crossengine;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.spark.sql.SparkSession;
+import org.junit.jupiter.api.*;
```
Comment (snazy) [inline nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java:25]: Don't use wildcard imports

Comment (snazy) [inline nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java:25]: Don't use wildcard imports

File: nqeit-iceberg-dremio/src/test/java/org/projectnessie/integtests/dremio/ITIcebergDremio.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.integtests.dremio;
+
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.*;
```
Comment (snazy) [inline nqeit-iceberg-dremio/src/test/java/org/projectnessie/integtests/dremio/ITIcebergDremio.java:21]: Don't use wildcard imports

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+})
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class ITCrossEngineExternalNessie {
+
+  private static SparkSession spark;
+  private static FlinkHelper flink;
+  private static DremioHelper dremioHelper;
+
+  @BeforeAll
+  public static void maybeSkipTests() {
```
Comment (snazy) [inline nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java:50]: Nothing guarantees that this method's called before `setupEngines`

File: nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java [hunk 4]
```diff
@@ -82,7 +82,10 @@ public URI getUri() {
 
   @Override
   public void close() throws Throwable {
-    deleteRecursively(tempDir);
+    if (tempDir != null) {
```
Comment (snazy) [inline nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java:85]: Not sure why this change is required for this PR.
Sure, being on the safe side is okay, but it seems that this temp-dir resource is added although it's not used?

**Review by snazy — CHANGES_REQUESTED** [review]
The approach could work. I'm mostly concerned about using rest-assuered, which is a _testing_ library, for this change - and also using manually crafted requests and response parsing, which is error-prone.

The README should have instructions about how to get the required information.

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import org.projectnessie.integtests.nessie.NessieTestsExtension;
+
+@ExtendWith({
+  IcebergSparkExtension.class,
+  IcebergFlinkExtension.class,
+  NessieTestsExtension.class,
+  IcebergDremioExtension.class
+})
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class ITCrossEngineExternalNessie {
```
Comment (snazy) [inline nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java:43]: The name of the test ("external Nessie") is likely wrong - I suspect, this test class is to test Dremio stuff with other engines?

File: settings.gradle.kts [hunk 6]
```diff
@@ -375,12 +376,16 @@ include("nqeit-iceberg-spark-extension")
...
 
+include("nqeit-iceberg-dremio-extension")
+
 fun includeProject(artifactId: String, projectDir: File) {
   include(artifactId)
   val p = project(":$artifactId")
   p.projectDir = projectDir
 }
 
+includeProject("nqeit-iceberg-dremio", file("nqeit-iceberg-dremio"))
```
Comment (snazy) [inline settings.gradle.kts:387]: A plain `include()` should be good

Comment (sagar-agarwal2404) [reply settings.gradle.kts]: Done.

File: nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java (hunk 4 above)
Comment (sagar-agarwal2404) [reply nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java:85]: https://github.com/projectnessie/query-engine-integration-tests/blob/80d275a9d8a2ba4832c7a464076de6d514e6c177/nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java#L51-L55

If the warehouse location is null then only we are setting the tempDir. In our case, we are using the warehouse location as S3, so if we do not handle this case, we will get error.

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java (hunk 3 above)
Comment (sagar-agarwal2404) [reply nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java]: The problem of using multiple @BeforeAll is been solved using interface. Added an interface class named Dremio. 

Doc states that "@BeforeAll methods from an interface will be executed before @BeforeAll methods in the class that implements the interface"
Reference used: https://junit.org/junit5/docs/5.4.2/api/org/junit/jupiter/api/BeforeAll.html

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java (hunk 5 above)
Comment (sagar-agarwal2404) [reply nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java]: I have changed the name to "ITCrossEngineDremio".

File: nqeit-iceberg-dremio-extension/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  compileOnly(libs.findbugs.jsr305)
+  compileOnly(libs.microprofile.openapi)
+  compileOnly(libs.jetbrains.annotations)
+  compileOnly(libs.errorprone.annotations)
+
+  compileOnly(platform(libs.junit.bom))
+  compileOnly(libs.junit.jupiter.engine)
+  compileOnly(libs.bundles.junit.testing)
+
+  implementation("io.rest-assured:rest-assured:$versionRestAsssured")
```
[the remark this answers is not in the document]
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/build.gradle.kts:38]: I will be using the Apache HttpComponents library. Is it fine to use this library? or if you could recommend some good library.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/IcebergDremioExtension.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Objects;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+public class IcebergDremioExtension implements ParameterResolver {
+
+  private static String readRequiredSystemProperty(String s) {
+    return Objects.requireNonNull(System.getProperty(s), "required system property not set: " + s);
```
[the remark this answers is not in the document]
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/IcebergDremioExtension.java:27]: i'd argue it is not ugly because the message is clearly describing what system property is missing.
we implemented the skipping as `Assumptions.assumeTrue(System.getProperty("dremio.url") != null);`  using the junit mechanism
but once that property is set, we want the tests to fail hard when the other properties are missing (instead of silently skipping the test).
this is why we use requireNonNull here.

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+  private static SparkSession spark;
```
[the remark this answers is not in the document]
Comment (sagar-agarwal2404) [reply nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java]: If we would have not used static fields then all the three parameters needs to be injected in every method along with other arguments . So to avoid that headache and the clarity of code, we have used the static fields instead of injecting them multiple times. 

for eg: In `assertSelectFromEngines()` method, we need to pass along 3 parameters all the time

File: nqeit-iceberg-dremio-extension/build.gradle.kts (hunk 7 above)
[the remark this answers is not in the document]
Comment (snazy) [reply nqeit-iceberg-dremio-extension/build.gradle.kts:38]: The point is, that you're plumbing the requests manually and parse the responses manually as well. I'd really prefer to use generated code based on the OpenAPI schema.

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java (hunk 3 above)
Comment (snazy) [reply nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java]: (From above) The new extension could easily do that skip, which is much cleaner than a @Before(All) or an interface.

File: nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java (hunk 4 above)
Comment (snazy) [reply nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java:85]: In this case, it would be preferable to let `getPath()` throw an `IllegalStateException`, if `path` is null.

Comment (XN137) [reply nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java:85]: when location is not null, then `tempDir` is not initialized, but this close method tries to clean it up unconditionally.
i am not sure how it is related to `getPath()´ but even if `getPath()` threw an exception as suggested, then the close method would still be wrong without this change, because it calls `deleteRecursively(null)`, no?

File: nqeit-iceberg-dremio-extension/build.gradle.kts (hunk 7 above)
[the remark this answers is not in the document]
Comment (XN137) [reply nqeit-iceberg-dremio-extension/build.gradle.kts:38]: can you link to the OpenAPI schema?
because we can not find it on https://docs.dremio.com/cloud/api/

also wondering:
is it really necessary to generate a full blown client from it?
as is, we are only making ~3 distinct/simple http calls in `DremioHelper` and they are well isolated internally.
we can rework/enhance this at a later point in time if really required.

i'd argue it is "good enough" as is and working on automatically running these tests would be higher priority than reworking this well isolated internal detail.

[the remark this answers is not in the document]
Comment (snazy) [reply nqeit-iceberg-dremio-extension/build.gradle.kts:38]: Not opposed to add it later. I assumed, that there is an OpenAPI schema available. Weird, that there are API examples, but not schema.

File: nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java (hunk 4 above)
Comment (sagar-agarwal2404) [reply nqeit-nessie-common/src/main/java/org/projectnessie/integtests/nessie/internal/AbstractStorageLocation.java:88]: Now, getPath() throws an IllegalStateException, if the path is null.

File: nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java (hunk 3 above)
Comment (sagar-agarwal2404) [reply nqeit-cross-engine/src/test/java/org/projectnessie/integtests/crossengine/ITCrossEngineExternalNessie.java]: Used extension to skip the tests if the `dremio-url` is not provided, and fail hard if any of the parameter is missing.

File: nqeit-iceberg-dremio-extension/build.gradle.kts (hunk 7 above)
[the remark this answers is not in the document]
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/build.gradle.kts]: Added instructions to README for getting the required properties and replaced the rest-assured library with http-client5.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  private List<List<Object>> parseQueryResult(String jobId) {
+
+    String url = baseUrl + "/v0/projects/" + projectId + "/job/" + jobId + "/results";
+    String result = executeHttpGet(url);
+
+    JSONObject jObject = new JSONObject(result);
+    int noOfRows = jObject.getInt("rowCount");
+    final JSONArray table_data = jObject.getJSONArray("rows");
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: Please avoid the `final` keyword - it has no runtime effect for variables + parameters

File: README.md [hunk 11]
```diff
@@ -222,6 +222,57 @@ result in failures.
...
+-Dnessie.client.ref=<ref_name> \
+-Ddremio.url=<dremio-url> \
+-Ddremio.token=<token> \
+-Ddremio.project-id=<project-id> \
+-Ddremio.catalog-name=<catalog-name> \
+-Dnessie.inttest.location.iceberg.warehouse==s3://my-bucket \
+-Dnessie.client.io-impl=org.apache.iceberg.aws.s3.S3FileIO"
+```
+Note:
+1. `dremio-url` and `dremio-token` should be set according to the following documentation: [dremio-url and dremio-token](https://docs.dremio.com/cloud/api/)
```
Comment (snazy) [inline README.md:272]: "should" -> "must"

File: nqeit-iceberg-dremio-extension/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  compileOnly(libs.findbugs.jsr305)
+  compileOnly(libs.microprofile.openapi)
+  compileOnly(libs.jetbrains.annotations)
+  compileOnly(libs.errorprone.annotations)
+
+  compileOnly(platform(libs.junit.bom))
+  compileOnly(libs.junit.jupiter.engine)
+  compileOnly(libs.bundles.junit.testing)
+
+  implementation("org.apache.httpcomponents.client5:httpclient5:5.2")
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/build.gradle.kts:38]: Dependencies should be declared in `gradle/libs.versions.toml`

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  private String createPayload(String query) {
+    String payload =
+        format("{ \"sql\": \"%s\", \"context\": [\"%s\", \"db\"] }", query, catalogName);
+    return payload;
+  }
+
+  private String executeHttpRequest(HttpUriRequestBase req) {
+    req.addHeader("Authorization", "Bearer " + token);
+    req.addHeader("Content-Type", "application/json");
+    try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:60]: Hm ... this creates a new http-client instance including a new connection pool for every request.
I'd say, either create the apache http client as a closeable resource for the test class in the Junit extension, or switch to the stupid & simple `URL.openConnection` (which would remove the dependency on the apache http client).

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+  private String submitQueryAndGetJobId(String query) {
+    String payload = createPayload(query);
+    String jobId = JsonPath.from(rest().body(payload).post("/sql").getBody().asString()).get("id");
+    return jobId;
+  }
+
+  private String awaitSqlJobResult(String query) {
+    String jobId = submitQueryAndGetJobId(query);
+    assertThat("COMPLETED").isEqualTo(waitForJobStatus(jobId));
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:117]: Better throw a checked `DremioRequestException` with a detailed message including the job ID + the status (and maybe whatever else is available). Tests may have to have `throws Exception` or so in their signature, but that's clearer IMO and allows adding tests that explicitly check for failures.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    return jobId;
+  }
+
+  public List<List<Object>> runSelectQuery(String query) {
+    String jobId = awaitSqlJobResult(query);
+    return parseQueryResult(jobId);
+  }
+
+  public void runInsertQuery(String query) {
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:126]: What's special about this one compared to `executeDmlStatement` that it deserves a special method?

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+      String result = executeHttpGet(url);
+      JSONObject jObject = new JSONObject(result);
+      jobState = jObject.getString("jobState");
+    }
+    return jobState;
+  }
+
+  private List<List<Object>> parseQueryResult(String jobId) {
+
+    String url = baseUrl + "/v0/projects/" + projectId + "/job/" + jobId + "/results";
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:92]: Please add a link to the web page describing the API

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  private String executeHttpGet(String url) {
+    HttpGet httpGet = new HttpGet(url);
+    return executeHttpRequest(httpGet);
+  }
+
+  private String waitForJobStatus(String jobId) {
+    String jobState = "RUNNING";
+    String url = baseUrl + "/v0/projects/" + projectId + "/job/" + jobId;
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:80]: Please add a link to the web page describing the API

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    for (int i = 0; i < noOfRows; i++) {
+      final JSONObject row = table_data.getJSONObject(i);
+      list.add(asList(row.getInt("id"), row.getString("val")));
+    }
+    return list;
+  }
+
+  private String submitQueryAndGetJobId(String query) {
+    String payload = createPayload(query);
+    String url = baseUrl + "/v0/projects/" + projectId + "/sql";
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:108]: Please add a link to the web page describing the API

**Review by snazy — COMMENTED** [review]
Thanks! Looks much better now!

Just not convinced that adding the apache http-client brings any benefit over URL/(Http)URLConnection. HC (with all its bells and whistles -> connection pool et al) would make sense, if it's instantiated (as a closeable resource) in the JUnit extension for the test class, so it can be reused.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 18 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:134]: done.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 17 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:104]: done

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 16 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:117]: done

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 15 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: Instead of using `executeDmlStatement()` and `runInsertQuery()` separately, we are now using a single method named `runQuery()`.

File: nqeit-iceberg-dremio-extension/build.gradle.kts (hunk 12 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/build.gradle.kts]: done.

File: README.md (hunk 11 above)
Comment (sagar-agarwal2404) [reply README.md]: done.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 10 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: done.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 14 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: Is it necessary to throw a checked `DremioQueryException` now? 
I've adjusted this spot to give more details about the job failure. however, apart from submitting an invalid SQL query, I couldn't think of any tests requiring the DremioRequestException so I thought we can add it later when we have some more test cases that require it. 
For the time being, if the job does not get completed, we display a fail message with Job Id, Job Status, and Query submitted.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 13 above)
Comment (snazy) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: Thanks

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 14 above)
Comment (snazy) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: Fair enough

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    return readResponse(con, url);
+  }
+
+  private String waitForJobStatus(String jobId) throws IOException {
+    // The doc for getting the job status for cloud is not there, but it is similar to software
+    // See docs: https://docs.dremio.com/software/rest-api/jobs/get-job/
+    String jobState = "RUNNING";
+    String url = baseUrl + "/v0/projects/" + projectId + "/job/" + jobId;
+    Set<String> finalJobStatesList = new HashSet<>(asList("COMPLETED", "FAILED", "CANCELED"));
+    while (!finalJobStatesList.contains(jobState)) {
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:107]: Hm. This can become an endless loop, and it's querying for the job status "as fast as possible".

Should have a timeout here, and sleep between iterations to not put too much pressure on the backend.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+      String result = performHttpRequest(url, null);
+      JsonNode node = parseJson(result);
+      jobState = node.get("jobState").textValue();
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:110]: Would be good to check for `FAILED` here, and throw an exception with the error message returned from the server.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+  DremioHelper(String projectId, String token, String baseUrl, String catalogName) {
+    this.projectId = projectId;
+    this.token = token;
+    this.baseUrl = baseUrl;
+    this.catalogName = catalogName;
+  }
+
+  private String createPayload(String query) {
+    String payload =
+        format("{ \"sql\": \"%s\", \"context\": [\"%s\", \"db\"] }", query, catalogName);
```
Comment (snazy) [inline nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:63]: This won't work, if the SQL query contains for example a `"`
Better construct the payload with JsonNode and let Jackson produce the string.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 19 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: done.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 20 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java:115]: done.

File: nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java (hunk 21 above)
Comment (sagar-agarwal2404) [reply nqeit-iceberg-dremio-extension/src/main/java/org/projectnessie/integtests/dremio/DremioHelper.java]: done.

**Review by snazy — APPROVED** [review]
Thanks! Approved + merging
